src/bt.rs

```diff
@@ -91,310 +97,290 @@ impl<V> Node<V> {
+    /// a reference to the node.
     ///
     /// # Returns
-    /// - `Ok(())` when the node is inserted at the end of the path.
-    /// - `Err(InsertNonEmptyNode(value))` when the subtree already contains a
+    /// - `Ok(NodeRef)` a reference is returned when the node is inserted
+    /// at the end of the path.
+    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
     /// value at the end of the path.
-    /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
+    /// - `Err(UnreachableNode(value))` when the end of the path is unreachable.
+    /// - `Err(InvalidNodeRef(value))` when the node reference is invalid.
     ///
     /// In the error cases, no insertion occurs and the value is returned to
     /// the caller of this function.
-    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        enum Ref<'a, V> {
-            This(&'a mut Node<V>),
-            Other(&'a mut Option<Box<Node<V>>>),
+    pub fn insert_at(
+        &mut self,
+        node_ref: NodeRef,
+        path: &Path,
+        value: V,
+    ) -> Result<NodeRef, BinaryTreeError<V>> {
+        if !(self.is_valid_node_ref(node_ref)
+            || (self.nodes.is_empty() && node_ref == NodeRef::ROOT))
+        {
+            return Err(BinaryTreeError::InvalidNodeRef(value));
         }
 
-        // Finds the node at the end of the path.
-        let mut node = Ref::This(self);
+        let new_index = self.nodes.len();
+        let mut node = &mut (!self.nodes.is_empty()).then_some(node_ref.0);
         for bit in path.iter() {
-            node = match node {
-                Ref::This(n) => Ref::Other(if !bit { &mut n.left } else { &mut n.right }),
-                Ref::Other(Some(n)) => Ref::Other(if !bit { &mut n.left } else { &mut n.right }),
-                Ref::Other(None) => {
-                    return Err(BinaryTreeError::UnreachableNode(value));
+            match *node {
+                None => return Err(BinaryTreeError::UnreachableNode(value)),
+                Some(next) => {
+                    let n = &mut self.nodes[next];
+                    node = if !bit { &mut n.left } else { &mut n.right }
                 }
-            };
+            }
         }
 
-        // Checks whether the node already has a value,
-        // If so, returns an error wrapping the value that could not be inserted.
-        // otherwise, a new node is created containing the value to be inserted.
-        match node {
-            Ref::This(_) | Ref::Other(Some(_)) => {
-                return Err(BinaryTreeError::InsertNonEmptyNode(value))
-            }
-            Ref::Other(empty_node) => {
-                *empty_node = Some(Box::new(Node::new(value)));
-            }
+        if node.is_some() {
+            return Err(BinaryTreeError::InsertNonEmptyNode(value));
         }
 
-        Ok(())
+        *node = Some(new_index);
+        self.nodes.push(Node::new(value));
```

Comment:
I'm probably just misreading the code, but I would expect that, somewhere in this function, we'd need to update the parent of the node we just inserted so that their node reference points to the node we just pushed. What am I missing?

Reply:
Yes, `node` is a mutable reference to the `Option<usize>` from the parent node's `left` or `right` link. The name `node` may be a bit misleading.